### PR TITLE
credocument: deal with loadDocument() failures

### PR DESCRIPTION
**Pull request by poire-z**
If cre:loadDocument() failed recognizing the document format (for example: a .epub in which /mimetype file is absent - crengine looks for this file in a zip to consider it a valid epub), koreader would still keep calling other methods on it, which would make crengine segfaults. With this, we can properly return an error
if document opening failed, so we never call other methods.
Will need a base bump for https://github.com/koreader/koreader-base/pull/582

### CODE REVIEW

**Comment by poire-z** [comment]
OK, these last failures should be no more once base is bumped.

File: frontend/document/credocument.lua [hunk 1]
```diff
@@ -109,22 +108,25 @@ function CreDocument:init()
     -- set fallback font face
     self._document:setStringProperty("crengine.font.fallback.face", self.fallback_font)
 
+    -- setting a default font before loading document actually do prevent
```
Comment (Frenzie) [inline frontend/document/credocument.lua]: Weird. Is it possible to add a couple of unit tests later to prevent regressions?

Comment (poire-z) [reply frontend/document/credocument.lua]: Not sure, this happens to me only with one non-free epub (Charles Juliet - Rencontres avec Samuel Beckett.epub , for my own futur reference).
I previously fixed this on the coverbrowser plugin side (#3293), as it's only on its side that crashes happen, but it needs now to be moved in credocument as delayed loadDocument is gone. It seems to me there is no performance penalty in setting 2 times a default font (this one, + the one from docsettings later), so it's hopefully fine.

**Comment by poire-z** [comment]
Mhhh, ok, may be i understand what happens:
Previously, loadDocument was called after all the onReadSettings, which would have set all the layout dimensions/fonts/size/interlines/margins...
With this, it is called early before all these are sets.
If i call loadDocument a second time after onReadSettings, it's all fine.
But loadDocument is expensive, so we can't allow 2 calls...
(but loadDocument shouldn't care much about all these settings, they should only be cared about by render()...)

But as it was at first commit, I fear it may screw up bookmarks and highlights, as the unit tests seem to indicate :(
Not sure what to do from there.
- Let it crash ?
- Let all the loadDocument where they were, and try to see what/how we can do when it returns false, without crashing
- in credocument.init, check if it has a sidecar file (which would mean it has bookmarks and custom settings, and that it **did** open previously without problem) and it if has, don't loadDocument in init ?

**Comment by poire-z** [comment]
Well... some thoughts about the 3 alternatives:

> Let all the loadDocument where they were, and try to see what/how we can do when it returns false, without crashing

That's not easy and clean, and adds complexity (coroutine, alternating between showReader and showFilemanager) for so little benefit (seen no issue about crashs because of unsupported file, so I guess it's not something that happen often - and it never happened to me before yesterday).

> in credocument.init, check if it has a sidecar file (which would mean it has bookmarks and custom settings, and that it did open previously without problem) and it if has, don't loadDocument in init ?

That would mean that second and next openings would have a different layout than the first one, so still the risks of messsing with bookmarks/highlights, and I guess unit tests would still fail.

> Let it crash ?

Seems the simpler: crash in readerui, with a more explicite error message (about document being not recognized by crengine) instead of a segfault.
We can still get the benefit of dealing better with the failure in in filemanagerbookinfo.lua and coverbrowser plugin.

**Comment by poire-z** [comment]
If you're ok with the final state of this, i'll squash and merge it to a proper commit message (i'd like to keep the experiment commits and CI results here, just in case I one day have to revisit it).

**Comment by Frenzie** [comment]
Sure! 👍

**Pull request merged**